db/model/Post.ts

```diff
@@ -72,3 +99,191 @@ export const getPostEnrichedBySlug = async (
+                    return subcategory.entries.some(
+                        (subCategoryEntry) => subCategoryEntry.slug === slug
+                    )
+                }
+            )
+        )
+    })
+}
+
+export const getPostsFromSnapshots = async (
+    postTypes: string[] = [WP_PostType.Post, WP_PostType.Page],
+    filterFunc?: FilterFnPostRestApi
+): Promise<PostRestApi[]> => {
+    const rawPosts: DbRawPost[] = (
+        await db.knexInstance().raw(
+            `
+                SELECT * FROM ${postsTable}
+                WHERE status = "publish"
+                AND type IN (?)
+                ORDER BY JSON_UNQUOTE(JSON_EXTRACT(wpApiSnapshot, '$.date')) DESC;
+            `,
+            [postTypes]
+        )
+    )[0]
+
+    const posts = rawPosts
+        .map(parsePostRow)
+        .map((p) => p.wpApiSnapshot)
+        .filter((p) => p !== null) as PostRestApi[]
+
+    // Published pages excluded from public views
+    const excludedSlugs = [BLOG_SLUG]
+
+    const filterConditions: Array<FilterFnPostRestApi> = [
+        (post): boolean => !excludedSlugs.includes(post.slug),
+        (post): boolean => !post.slug.endsWith("-country-profile"),
+    ]
+    if (filterFunc) filterConditions.push(filterFunc)
+
+    return posts.filter((post) => filterConditions.every((c) => c(post)))
+}
+
+export const getPostRelatedCharts = async (
+    postId: number
+): Promise<RelatedChart[]> =>
+    db.queryMysql(`
+        SELECT DISTINCT
+            charts.config->>"$.slug" AS slug,
+            charts.config->>"$.title" AS title,
+            charts.config->>"$.variantName" AS variantName,
+            chart_tags.keyChartLevel
+        FROM charts
+        INNER JOIN chart_tags ON charts.id=chart_tags.chartId
+        INNER JOIN post_tags ON chart_tags.tagId=post_tags.tag_id
+        WHERE post_tags.post_id=${postId}
+        AND charts.config->>"$.isPublished" = "true"
+        ORDER BY title ASC
+    `)
+
+export const getFullPost = async (
+    postApi: PostRestApi,
+    excludeContent?: boolean
+): Promise<FullPost> => ({
+    id: postApi.id,
+    type: postApi.type,
+    slug: postApi.slug,
+    path: postApi.slug, // kept for transitioning between legacy BPES (blog post as entry section) and future hierarchical paths
+    title: decodeHTML(postApi.title.rendered),
+    date: new Date(postApi.date_gmt),
+    modifiedDate: new Date(postApi.modified_gmt),
+    authors: postApi.authors_name || [],
+    content: excludeContent ? "" : postApi.content.rendered,
+    excerpt: decodeHTML(postApi.excerpt.rendered),
+    imageUrl: `${BAKED_BASE_URL}${
+        postApi.featured_media_paths.medium_large ?? "/default-thumbnail.jpg"
+    }`,
+    thumbnailUrl: `${BAKED_BASE_URL}${
+        postApi.featured_media_paths?.thumbnail ?? "/default-thumbnail.jpg"
+    }`,
+    imageId: postApi.featured_media,
+    relatedCharts:
+        postApi.type === "page"
+            ? await getPostRelatedCharts(postApi.id)
```

Comment:
I think this is not a good query pattern - it looks like quite a few code paths need the result of getFullPost for many posts, but by doing it this way we have to run one query against MySQL for every post. I think it would be much better to have two flavours of getPostsFromSnapshots, one like it is at the moment, and one that also queries for the related chart information (doing a join and then a group by). This second query would mean that getting full posts for all posts can be done with a single query against the DB. Admittedly, this one query would then become quite complex, but I think this is still worth it. I'm happy to help with writing the refined query, just let me know.

Reply:
that makes sense, thanks for calling this out. Issue created here: https://github.com/owid/owid-grapher/issues/3192, and tracked here: #3097